docs/source/data.rst

```diff
@@ -2,11 +2,219 @@ torch.utils.data
+gradient decent (SGD), a :class:`~torch.utils.data.Sampler` could randomly
+permute a list of indices and yield each one at a time, or yield a small number
+of them for mini-batch SGD.
+
+`Data Loading Strategies`_ section talks about how to use a :class:`~torch.utils.data.Sampler`
+with a :class:`~torch.utils.data.DataLoader`.
+
+Data Loading Strategies
+-----------------------
+
+:class:`~torch.utils.data.DataLoader` constructor receives a
+:attr:`dataset` object as its first argument. Based on the other provided
+arguments, a :class:`~torch.utils.data.DataLoader` operates in one of three
+following modes:
+
+* `Batched loading from a map-style dataset (default)`_
+
+* `Loading individual members of a map-style dataset`_
+
+* `Loading from an iterable-style dataset`_
+
+Batched loading from a map-style dataset (default)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This is the most common case, and corresponds to fetching a minibatch of
+data and collating them into batched Tensors, i.e., Tensors with one dimension
+being the batch dimension (usually the first). Two combinations of arguments
+that starts this mode are:
+
+* (default) Using arguments :attr:`batch_size`, :attr:`shuffle`,
+  :attr:`sampler`, and :attr:`drop_last` to specify the batch indices sampling
+  behavior.
+
+  With the default arguments, a :class:`~torch.utils.data.DataLoader`
+  loads data as batches of size ``1`` with indices sampled without replacement.
+
+* Setting argument :attr:`batch_sampler` to a custom sampler returning a list
+  of indices at each time, representing the indices for a batch.
+
+After fetching a list of samples using the indices from sampler, :attr:`collate_fn`
+is used to collate the list of samples into batched Tensors. Users may use
+customized :attr:`collate_fn` to achieve custom batching, e.g., along a
+dimension other than the first.
+
+The behavior of this mode is roughly equivalent with::
+
+    for indices in batch_sampler:
+        yield collate_fn([dataset[i] for i in indices])
+
+Loading individual members of a map-style dataset
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Sometimes it is cheaper to directly load batched data (e.g., bulk reads from a
+database or a remote server), or your program is designed for working on
+individual samples.  In such cases, it's probably better to not use the above
+loading strategy (where :attr:`collate_fn` is used to collate the samples), but
+let the data loader directly return each member of the :attr:`dataset` object.
+
+To start this mode, simply set ``batch_size=None``.  If :attr:`sampler` is set,
+it will be used to sample the indices/keys to :attr:`dataset`. Otherwise, a
+sampler will be constructed according to :attr:`shuffle` argument.
+
+The behavior of this mode is roughly equivalent with::
+
+    for index in sampler:
+        yield convert_fn(dataset[index])
+
+
+Loading from an iterable-style dataset
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+
```

Comment:
It would be nice to have an example here so that people not familiar with the `utils.data` internals can still discuss the API and usage scenarios.

Reply:
Thanks for the suggestion. I agree. This part isn't finished yet. :)

Reply:
no problem, take the time, its an important feature ;-)

I just thought it would be nice to do these docs in an early stage so that "we users" could comment on the API